api/client/build.go

```diff
@@ -315,3 +277,295 @@ func (cli *DockerCli) CmdBuild(args ...string) error {
 	}
 	return err
 }
+
```

Comment:
The following 6 functions should actually make implementing the client side builder much easier because they handle the work of preparing the context directory and locating the Dockerfile. Those will not be going away once we have a client-side builder. All of that refactoring work was done in a separate commit. It was necessary to do this now so that we get then have an obvious point in the `CmdBuild` method above to perform what we need to do now with the Dockerfile.

The second commit adds the two functions that are at the end of this file (the one that parses and rewrites the `FROM` lines in a Dockerfile, and the one that modifies the tar stream). These are the ones that are *just* for this PR/feature and will be deleted once we have a client-side builder.